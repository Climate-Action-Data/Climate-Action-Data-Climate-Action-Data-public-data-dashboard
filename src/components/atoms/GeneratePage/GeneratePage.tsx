/* eslint-disable no-magic-numbers */
import React from 'react'
import { Box, Button } from '@chakra-ui/react'

interface GeneratePageProps {
  currentPage: number
  totalPages: number
  onPageClick: (page: number) => void
}

const DEFAULT_PAGES_TO_SHOW = 4

export const GeneratePage = (props: GeneratePageProps) => {
  const { currentPage, totalPages, onPageClick } = props
  const renderPageNumbers = () => {
    const pageNumbers: React.JSX.Element[] = []

    // We should display all the pages if there are 4 or fewer pages
    if (totalPages <= DEFAULT_PAGES_TO_SHOW) {
      for (let i = 1; i <= totalPages; i++) {
        pageNumbers.push(
          <Button variant="lightGrayRound32" key={i} fontWeight={currentPage === i ? `bold` : `500`} mx={1} onClick={() => onPageClick(i)}>
            {i}
          </Button>,
        )
      }
    } else {
      // Display ellipsis if there are more than 4 pages and current page is not close to the start
      if (currentPage > 2) {
        pageNumbers.push(
          <Button key={1} variant="lightGrayRound32" mx={1} onClick={() => onPageClick(1)}>
            1
          </Button>,
        )
        if (currentPage > 3) {
          pageNumbers.push(
            <Box flex={1} key="ellipsis1">
              ...
            </Box>,
          )
        }
      }

      // Display current page and nearby pages
      for (let i = currentPage - 1; i <= currentPage + 1; i++) {
        if (i > 0 && i <= totalPages) {
          pageNumbers.push(
            <Button key={i} fontWeight={currentPage === i ? `bold` : `500`} variant="lightGrayRound32" mx={1} onClick={() => onPageClick(i)}>
              {i}
            </Button>,
          )
        }
      }

      // Display ellipsis if there are more than 4 pages and current page is not close to the end
      if (currentPage < totalPages - 1) {
        if (currentPage < totalPages - 2) {
          pageNumbers.push(
            <Box data-testid="ellipsis" flex={1} key="ellipsis2">
              ...
            </Box>,
          )
        }
        pageNumbers.push(
          <Button data-testid="pagination-last-page" key={totalPages} variant="lightGrayRound32" mx={1} onClick={() => onPageClick(totalPages)}>
            {totalPages}
          </Button>,
        )
      }
    }

    return pageNumbers
  }

  return <React.Fragment>{renderPageNumbers()}</React.Fragment>
}
