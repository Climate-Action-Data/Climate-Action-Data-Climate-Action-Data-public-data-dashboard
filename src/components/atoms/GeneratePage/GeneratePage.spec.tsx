import { fireEvent, render, screen } from '@testing-library/react'
import { GeneratePage } from './GeneratePage'

const DEFAULT_CURRENT_PAGE = 1
const DEFAULT_TOTAL_PAGES = 6
const DEFAULT_TOTAL_PAGES_XS = 2

it(`renders correctly`, () => {
  const onPageChange = jest.fn()
  const { container } = render(<GeneratePage currentPage={DEFAULT_CURRENT_PAGE} totalPages={DEFAULT_TOTAL_PAGES} onPageClick={(page) => onPageChange(page)} />)

  expect(container).toMatchSnapshot()
})

it(`renders correctly with xs pages`, () => {
  const onPageChange = jest.fn()

  const { container } = render(<GeneratePage currentPage={DEFAULT_CURRENT_PAGE} totalPages={DEFAULT_TOTAL_PAGES_XS} onPageClick={(page) => onPageChange(page)} />)

  expect(container).toMatchSnapshot()
})

it(`renders correctly with xs pages and click`, () => {
  const onPageChange = jest.fn()
  const { container } = render(<GeneratePage currentPage={DEFAULT_CURRENT_PAGE} totalPages={DEFAULT_TOTAL_PAGES_XS} onPageClick={(page) => onPageChange(page)} />)

  const inputElement = screen.getByText(`${DEFAULT_CURRENT_PAGE + 1}`)
  fireEvent.click(inputElement)

  expect(onPageChange).toHaveBeenCalled()
  expect(container).toMatchSnapshot()
})

it(`renders correctly and click next`, () => {
  const onPageChange = jest.fn()
  const { container } = render(<GeneratePage currentPage={DEFAULT_CURRENT_PAGE} totalPages={DEFAULT_TOTAL_PAGES} onPageClick={(page) => onPageChange(page)} />)

  const inputElement = screen.getByText(`1`)
  fireEvent.click(inputElement)

  expect(container).toMatchSnapshot()
  expect(onPageChange).toHaveBeenCalled()
})

it(`renders correctly and click down from max`, () => {
  const onPageChange = jest.fn()
  const { container } = render(<GeneratePage currentPage={DEFAULT_TOTAL_PAGES} totalPages={DEFAULT_TOTAL_PAGES} onPageClick={(page) => onPageChange(page)} />)

  const inputElement = screen.getByText(`${DEFAULT_TOTAL_PAGES - 1}`)
  fireEvent.click(inputElement)

  expect(container).toMatchSnapshot()
  expect(onPageChange).toHaveBeenCalled()
})

it(`renders correctly and click first page`, () => {
  const onPageChange = jest.fn()
  const { container } = render(<GeneratePage currentPage={DEFAULT_TOTAL_PAGES} totalPages={DEFAULT_TOTAL_PAGES} onPageClick={(page) => onPageChange(page)} />)

  const inputElement = screen.getByText(`1`)
  fireEvent.click(inputElement)

  expect(onPageChange).toHaveBeenCalled()
  expect(container).toMatchSnapshot()
})
