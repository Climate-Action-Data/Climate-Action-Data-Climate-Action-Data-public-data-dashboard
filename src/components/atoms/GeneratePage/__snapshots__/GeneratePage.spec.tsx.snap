// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<div>
  <button
    class="chakra-button css-6iqjvk"
    type="button"
  >
    1
  </button>
  <button
    class="chakra-button css-1hwtise"
    type="button"
  >
    2
  </button>
  <div
    class="css-1rr4qq7"
    data-testid="ellipsis"
  >
    ...
  </div>
  <button
    class="chakra-button css-1ki210r"
    data-testid="pagination-last-page"
    type="button"
  >
    6
  </button>
</div>
`;

exports[`renders correctly and click down from max 1`] = `
<div>
  <button
    class="chakra-button css-1ki210r"
    type="button"
  >
    1
  </button>
  <div
    class="css-1rr4qq7"
  >
    ...
  </div>
  <button
    class="chakra-button css-1hwtise"
    type="button"
  >
    5
  </button>
  <button
    class="chakra-button css-6iqjvk"
    type="button"
  >
    6
  </button>
</div>
`;

exports[`renders correctly and click first page 1`] = `
<div>
  <button
    class="chakra-button css-1ki210r"
    type="button"
  >
    1
  </button>
  <div
    class="css-1rr4qq7"
  >
    ...
  </div>
  <button
    class="chakra-button css-1hwtise"
    type="button"
  >
    5
  </button>
  <button
    class="chakra-button css-6iqjvk"
    type="button"
  >
    6
  </button>
</div>
`;

exports[`renders correctly and click next 1`] = `
<div>
  <button
    class="chakra-button css-6iqjvk"
    type="button"
  >
    1
  </button>
  <button
    class="chakra-button css-1hwtise"
    type="button"
  >
    2
  </button>
  <div
    class="css-1rr4qq7"
    data-testid="ellipsis"
  >
    ...
  </div>
  <button
    class="chakra-button css-1ki210r"
    data-testid="pagination-last-page"
    type="button"
  >
    6
  </button>
</div>
`;

exports[`renders correctly with xs pages 1`] = `
<div>
  <button
    class="chakra-button css-6iqjvk"
    type="button"
  >
    1
  </button>
  <button
    class="chakra-button css-1hwtise"
    type="button"
  >
    2
  </button>
</div>
`;

exports[`renders correctly with xs pages and click 1`] = `
<div>
  <button
    class="chakra-button css-6iqjvk"
    type="button"
  >
    1
  </button>
  <button
    class="chakra-button css-1hwtise"
    type="button"
  >
    2
  </button>
</div>
`;
